allow rate-limiting on data nodes (for shards.tolerant=true)

magibney opened the pull request:
this is a quick-and-dirty hack that will work for our usage, but this should be reconsidered and something more general committed upstream.

Honestly it doesn't make sense to hardcode handling of this at the level of RateLimitManager. Individual rate limiters should be specified as plugins, with the context necessary to make their own decisions.

magibney committed:
allow rate-limiting on data nodes (for shards.tolerant=true)

this is a quick-and-dirty hack that will work for our
usage, but this should be reconsidered and something
more general committed upstream.

Honestly it doesn't make sense to hardcode handling of
this at the level of RateLimitManager. Individual rate
limiters should be specified as plugins, with the
context necessary to make their own decisions.

magibney commented:
opening this against fs/branch_9_3 for quick evaluation

hiteshk25 commented on solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java:
```diff
@@ -586,6 +587,17 @@ private Request makeRequest(SolrRequest<?> solrRequest, String collection)
+    SolrRequest.SolrClientContext context = getContext();
+    req.header(CommonParams.SOLR_REQUEST_CONTEXT_PARAM, context.toString());
+    if (context == SolrRequest.SolrClientContext.CLIENT
+        || solrRequest.getParams().getBool(ShardParams.SHARDS_TOLERANT, false)) {
```
Whole purpose of ratelimiter to limit the resources on all nodes. Not sure we really need to make any special case here.

magibney replied:
The reason for differentiating between `shards.tolarant=true` vs. `shards.tolerant=false` is explained in the comment immediately below:
```java
      // NOTE: if `shards.tolerant=false`, do _not_ set the `Solr-Request-Type` header, because we
      // could end up doing a lot of extra work at the cluster level, retrying requests that may
      // only have failed to obtain a ratelimit permit on a single shard.
```

magibney replied:
For our case in particular, practically we actually do want to avoid failing any requests that are `shards.tolerant=false`. Notably these are also the most likely requests to be retried on failure, so if we end up repeatedly executing requests on all nodes, only to repeatedly fail because of 1 struggling node (for example), that could easily cause load to increase on the other nodes to the point where the problem spreads to the entire cluster.

hiteshk25 replied:
I guess these are two different problems. Rate limiting should be agnostic to any parameters. 
if we see any issue with `shards.tolerant=false` or single node, then we need to track that in that context. **We don't want to increase the load on other node same time.**

magibney replied:
I'm not sure what you're suggesting to do here. If we see an issue with `shards.tolerant=false` on a single node, we will _already_ have increased the load on other nodes (requests to nodes are issued in parallel), and if the top-level request fails due to the one node rate-limiting, then the client is likely to retry, increasing the load on the cluster overall (the exact situation that we both agree we want to avoid).

> Rate limiting should be agnostic to any parameters

Why do you say this? I think the status quo (evaluate rate limiting only on the coordinator node) is due to the potential for rate-limiting evaluation on data-nodes to amplify request load. So if we really want rate limiting to be agnostic to any parameters, then I think rate-limiting on data nodes must be avoided entirely (due to the request amplification issue).

hiteshk25 replied:
Let's chat on this.

magibney committed:
support sysprop-based toggling of ratelimiting behavior wrt shards.tolerant

adds `solr.dataNodeAllowRateLimitAlways` param (true/false,
default false. If false, data node rateLimiting is only
applied when `shards.tolerant=true`, otherwise (if `true`),
rateLimiting may always be applied on data node for distrib
requests, even if `shards.tolerant=false`.

hiteshk25 commented on solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java:
```diff
@@ -592,6 +601,20 @@ private void setBasicAuthHeader(SolrRequest<?> solrRequest, Request req) {
+      // only have failed to obtain a ratelimit permit on a single shard.
+      req.headers(
+          headers ->
+              headers.put(CommonParams.SOLR_REQUEST_TYPE_PARAM, solrRequest.getRequestType()));
```
In our case, we need to pass following header as well. Will that work as it is?
```
  public static final String SOLR_REQUEST_PRIORITY_PARAM = "Solr-Request-Priority";
```

magibney replied:
ooh good call. I'll add that.

We probably in the near future want to come up with a more generic way to do this, rather than hardcode expectations around these things 🤔 ... but for now this is fine.

magibney committed:
move logic into header propagation SearchHandler

magibney committed:
upgrade github workflow `actions/cache` v2->v4

https://github.blog/changelog/2024-12-05-notice-of-upcoming-releases-and-breaking-changes-for-github-actions/#actions-cache-v1-v2-and-actions-toolkit-cache-package-closing-down

magibney committed:
add Override annotation

magibney committed:
fix bug in shards.tolerant parsing

magibney commented:
@hiteshk25 this should be functional now (pending test, forthcoming). You [have to add](https://solr.apache.org/guide/solr/latest/configuration-guide/requestdispatcher.html#requestparsers-element) the `addHttpRequestToContext="true"` attribute to `solrconfig.xml` in order for this to work.
```xml
    <requestDispatcher handleSelect="false" >
        <requestParsers multipartUploadLimitInKB="[value]" formdataUploadLimitInKB="[value]" addHttpRequestToContext="true" />
    </requestDispatcher>
```

magibney committed:
Merge remote-tracking branch 'origin/fs/branch_9_7' into michaelgibney/data-node-ratelimiting

magibney committed:
Merge remote-tracking branch 'origin/fs/branch_9_7' into michaelgibney/data-node-ratelimiting

hiteshk25 commented:
> @hiteshk25 this should be functional now (pending test, forthcoming). You [have to add](https://solr.apache.org/guide/solr/latest/configuration-guide/requestdispatcher.html#requestparsers-element) the `addHttpRequestToContext="true"` attribute to `solrconfig.xml` in order for this to work.
> 
> ```
>     <requestDispatcher handleSelect="false" >
>         <requestParsers multipartUploadLimitInKB="[value]" formdataUploadLimitInKB="[value]" addHttpRequestToContext="true" />
>     </requestDispatcher>
> ```

Wondering what will happen for other handlers export/stream etc?

hiteshk25 reviewed (approved):
LGTM
